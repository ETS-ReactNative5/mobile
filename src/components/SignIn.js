import React from 'react'
import {
  Alert,
  Linking,
  Text,
  TouchableOpacity,
  View
} from 'react-native'
import EStyleSheet from 'react-native-extended-stylesheet'
import { signIn } from '../actions/index'
import { connect } from 'react-redux';
import Button from './Button'
import Input from './Input'
import NavBar from './NavBar'
import OverlaySpinner from './OverlaySpinner'
import StyledText from './StyledText'

const mapStateToProps = (state) => ({
  isFetching: state.isFetching,
  errorMessage: state.errorMessage,
  isConnected: state.isConnected,
})

const mapDispatchToProps = (dispatch) => ({
  signIn(login, password) {
    dispatch(signIn(login, password))
  },
})

class SignIn extends React.Component {
  constructor(props) {
    super(props)
    this.state = {login: '', password: ''}
    this.handleRegistration = this.handleRegistration.bind(this)
    this.handleResetPassword = this.handleResetPassword.bind(this)
    this.handleSignIn = this.handleSignIn.bind(this)
  }

  handleRegistration() {
    const zurl='https://www.zooniverse.org/accounts/register'
    Linking.canOpenURL(zurl).then(supported => {
      if (supported) {
        Linking.openURL(zurl);
      } else {
        Alert.alert(
          'Error',
          'Sorry, but it looks like you are unable to open your default browser.',
        )
      }
    });
  }

  handleResetPassword() {
    const zurl='http://zooniverse.org/reset-password'
    Linking.canOpenURL(zurl).then(supported => {
      if (supported) {
        Linking.openURL(zurl);
      } else {
        Alert.alert(
          'Error',
          'Sorry, but it looks like you are unable to reset your password in your default browser.',
        )
      }
    });
  }

  handleSignIn() {
    //prevent red screen of death thrown by a console.error in javascript-client
    /* eslint-disable no-console */
    console.reportErrorsAsExceptions = false
    this.props.signIn(this.state.login, this.state.password)
  }

  render() {
    const signInDisabled = ( (this.state.login === '') || (this.state.password ===  '') ? true : false )

    return (
      <View style={styles.container}>
        <NavBar showLogo={true} showDrawer={false} />
        <View style={styles.signInContainer}>
          <Button
            handlePress={this.handleRegistration}
            text={'REGISTER FOR ACCOUNT'} />
          <View style={styles.lined}>
            <View style={styles.lineThrough} />
            <Text style={styles.centerText}>OR</Text>
            <View style={styles.lineThrough} />
          </View>

          <StyledText
            textStyle={'headerText'}
            text={'SIGN IN'} />
          <Input
            labelText={'Username or Email Address'}
            handleOnChangeText={(login) => this.setState({login})} />
          <Input
            labelText={'Password'}
            passwordField={true}
            handleOnChangeText={(password) => this.setState({password})} />
          <StyledText
            textStyle={'errorMessage'}
            text={ this.props.errorMessage } />
          <TouchableOpacity
            onPress={this.handleResetPassword}>
            <StyledText
              textStyle={'link'}
              text={ 'Forget your password?' } />
          </TouchableOpacity>

          <Button
            handlePress={this.handleSignIn}
            disabled={signInDisabled}
            text={'Sign In'} />
        </View>
        { this.props.isFetching ? <OverlaySpinner /> : null }
      </View>
    );
  }
}

const styles = EStyleSheet.create({
  container: {
    flex: 1,
    paddingTop: 60,
  },
  signInContainer: {
    backgroundColor: 'white',
    margin: 30,
    paddingLeft: 15,
    paddingRight: 15,
  },
  lined: {
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  lineThrough: {
    flex: 1,
    height: 1,
    backgroundColor: 'black'
  },
  centerText: {
    margin: 20
  }
});

SignIn.propTypes = {
  isFetching: React.PropTypes.bool.isRequired,
  signIn: React.PropTypes.func.isRequired,
  errorMessage: React.PropTypes.string
}

export default connect(mapStateToProps, mapDispatchToProps)(SignIn)
