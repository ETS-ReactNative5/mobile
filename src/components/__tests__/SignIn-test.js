import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'
import { SignIn } from '../SignIn'

it('renders correctly', () => {
  const tree = renderer.create(
    <SignIn isConnected={true} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders error message', () => {
  const tree = renderer.create(
    <SignIn isConnected={true} errorMessage={'does not compute'} />
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
