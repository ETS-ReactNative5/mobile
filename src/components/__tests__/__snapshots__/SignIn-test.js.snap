exports[`test renders correctly 1`] = `
<View
  style={undefined}>
  <ScrollView>
    <View
      style={undefined}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        SIGN IN
      </Text>
      <View
        style={undefined}>
        <View
          style={undefined}>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}>
            Username or Email Address
          </Text>
        </View>
        <View
          style={undefined}>
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            keyboardType="default"
            onChangeText={[Function]}
            returnKeyType="done"
            secureTextEntry={false}
            style={undefined}
            underlineColorAndroid="white" />
        </View>
      </View>
      <View
        style={undefined}>
        <View
          style={undefined}>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}>
            Password
          </Text>
        </View>
        <View
          style={undefined}>
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            keyboardType="default"
            onChangeText={[Function]}
            returnKeyType="done"
            secureTextEntry={true}
            style={undefined}
            underlineColorAndroid="white" />
        </View>
      </View>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={undefined}>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              undefined,
              undefined,
            ]
          }>
          Forget your password?
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        activeOpacity={0.5}
        disabled={true}
        onPress={[Function]}
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          Sign In
        </Text>
      </TouchableOpacity>
      <View
        style={undefined}>
        <View
          style={undefined} />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          OR
        </Text>
        <View
          style={undefined} />
      </View>
      <TouchableOpacity
        activeOpacity={0.5}
        disabled={false}
        onPress={[Function]}
        style={undefined}>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          Continue without signing in
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        activeOpacity={0.5}
        disabled={false}
        onPress={[Function]}
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          Register for account
        </Text>
      </TouchableOpacity>
    </View>
  </ScrollView>
</View>
`;

exports[`test renders error message 1`] = `
<View
  style={undefined}>
  <ScrollView>
    <View
      style={undefined}>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        SIGN IN
      </Text>
      <View
        style={undefined}>
        <View
          style={undefined}>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}>
            Username or Email Address
          </Text>
        </View>
        <View
          style={undefined}>
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            keyboardType="default"
            onChangeText={[Function]}
            returnKeyType="done"
            secureTextEntry={false}
            style={undefined}
            underlineColorAndroid="white" />
        </View>
      </View>
      <View
        style={undefined}>
        <View
          style={undefined}>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={undefined}>
            Password
          </Text>
        </View>
        <View
          style={undefined}>
          <TextInput
            autoCapitalize="none"
            autoCorrect={false}
            keyboardType="default"
            onChangeText={[Function]}
            returnKeyType="done"
            secureTextEntry={true}
            style={undefined}
            underlineColorAndroid="white" />
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        does not compute
      </Text>
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={undefined}>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              undefined,
              undefined,
            ]
          }>
          Forget your password?
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        activeOpacity={0.5}
        disabled={true}
        onPress={[Function]}
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          Sign In
        </Text>
      </TouchableOpacity>
      <View
        style={undefined}>
        <View
          style={undefined} />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          OR
        </Text>
        <View
          style={undefined} />
      </View>
      <TouchableOpacity
        activeOpacity={0.5}
        disabled={false}
        onPress={[Function]}
        style={undefined}>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          Continue without signing in
        </Text>
      </TouchableOpacity>
      <TouchableOpacity
        activeOpacity={0.5}
        disabled={false}
        onPress={[Function]}
        style={
          Array [
            undefined,
            undefined,
          ]
        }>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={undefined}>
          Register for account
        </Text>
      </TouchableOpacity>
    </View>
  </ScrollView>
</View>
`;
